aws_health_to_slack/program.py

```diff
@@ -0,0 +1,115 @@
+"""Announce AWS health events in Slack, based on resource ownership in Google Sheet.
+
+Documentation:
+https://docs.aws.amazon.com/health/
+https://aws.amazon.com/blogs/mt/tag/aws-health-api/
+"""
+
+from datetime import UTC, datetime, timedelta
+import json
+import os
+import re
+
+import autokitteh
+from autokitteh.aws import boto3_client
+from autokitteh.google import google_sheets_client
+from autokitteh.slack import slack_client
+
+
+url = os.getenv("GOOGLE_SHEET_URL")
+
+
+def on_every_minute(_):
+    """Workflow's entry-point."""
+    slack_channels = _read_google_sheet()
+    events = _aws_health_events()
+    events_by_arn = {event["arn"]: event for event in events}
+
+    for entity in _affected_aws_entities(events):
+        project = entity.get("tags", {}).get("project")
+        if not project:
+            print(f"Error: AWS entity without project tag: {entity}")
+            continue
+
+        channel = slack_channels.get(project)
+        affecting_events = [events_by_arn[arn] for arn in entity["eventArns"]]
+        _post_slack_message(project, channel, entity, affecting_events)
+
+
+@autokitteh.activity
+def _read_google_sheet() -> dict[str, str]:
+    """Read mapping of project tags to Slack channels from Google Sheet."""
+    match = re.match(r"(.*/d/)?([\w-]{20,})", url)
+    if not match:
+        raise ValueError(f"Error: invalid Google Sheet ID in URL {url!r}")
+
+    id = match.group(2)
+    sheets = google_sheets_client("google_sheets_connection").spreadsheets().values()
+    rows = sheets.get(spreadsheetId=id, range="A:B").execute().get("values", [])
+    return {row[0].strip(): row[1].strip() for row in rows}
+
+
+@autokitteh.activity
+def _aws_health_events() -> list[dict]:
+    """List all recent AWS Health events.
+
+    API Reference:
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/health/client/describe_events.html
+    """
+    try:
+        aws = boto3_client("aws_connection", "health")
+        dt = datetime.now(UTC) - timedelta(minutes=1)
+        filter = {"lastUpdatedTime": [{"from": dt}]}
+        # Possible alternative: describe_events_for_organization.
```

Comment:
I'm not sure I understand what this comment's intended purpose is. Is this a comment for reviewers to consider a different name for the function below? Or is it meant for the user?

Reply:
`describe_events` and `describe_events_for_organization` are similar-but-different functions in `boto3` (the AWS SDK).

The project is currently using `describe_events` to fetch events for a single AWS account, but with some more setup code we could easily replace it with `describe_events_for_organization` to fetch events for **multiple** AWS accounts.

This would be more useful for large-scale users (i.e. mid-to-large companies), but also more complicated and less appealing for people who want to read the code.

So I guess this is somewhere between a self-reminder and a suggestion for "the interested reader", but not an actual TODO.

Do you think moving this comment to the function's docstring as a clear full sentence would be better?

Reply:
For an inexperienced person yes. I think a complete sentence would make it much more clear.
I think that comment would only make sense to someone who knows those two functions. Which may be ok?

Reply:
Done